File: Java/Data Structures & Algorithms Module - Java/Week10/BSTAttempt5/BST.java
/*
BST.java
Billy Cussen
09/12/2020
*/

public class BST<T extends Comparable<T>> implements BSTInterface<T>{

    BSTNode<T> root;

    public boolean isEmpty(){
        return this.root == null;
    }

    public int size(){
        if(isEmpty()){
            return 0;
        }
        return sizeHelper(this.root);
    }

    private int sizeHelper(BSTNode<T> current){
        if(current==null){
            return 0;
        }
        return 1 + sizeHelper(current.left)+sizeHelper(current.right);
    }

    public void insert(T element){
        if(isEmpty()){
            this.root = new BSTNode<T>(element);
        } 
        insertHelper(this.root, element);
    }

    private void insertHelper(BSTNode<T> current, T element){
        if(element.compareTo(current.element)==-1){
            if(current.left == null){
                current.left = new BSTNode<T>(element);
            } else{
                insertHelper(current.left, element);
            }
        } else{
            if(current.right==null){
                current.right = new BSTNode<T>(element);
            } else{
                insertHelper(current.right, element);
            }
        }
    }

    public void inOrder(){
        if(isEmpty()){
            throw new NullPointerException("Can't do in order Traversal, BST is Empty!");
        }
        System.out.println("***BST IN ORDER TRAVERSAL***");
        inOrderHelper(this.root);
        System.out.println("***END***");
    }

    private void inOrderHelper(BSTNode<T> current){
        if(current==null){
            return;
        }
        inOrderHelper(current.left);
        System.out.println(current);
        inOrderHelper(current.right);
    }

    public T max(){
        if(isEmpty()){
            throw new NullPointerException("Can't get Max, BST is Empty!");
        }
        return maxHelper(this.root);
    }

    private T maxHelper(BSTNode<T> current){
        if(current.right==null){
            return current.element;
        } else{
            return maxHelper(current.right);
        }
    }

    public T min(){
        if(isEmpty()){
            throw new NullPointerException("Can't get Min, BST is Empty!");
        }
        return minHelper(this.root);
    }

    private T minHelper(BSTNode<T> current){
        if(current.left == null){
            return current.element;
        } else{
            return minHelper(current.left);
        }
    }

    public int depth(){
        if(isEmpty()){
            return 0;
        }
        return depthHelper(this.root);
    }

    private int depthHelper(BSTNode<T> current){
        if(current==null){
            return 0;
        }

        int left = depthHelper(current.left);
        int right = depthHelper(current.right);

        if(left > right){
            return 1 + depthHelper(current.left);
        } else{
            return 1 + depthHelper(current.right);
        }
    }

    public boolean contains(T element){
        if(isEmpty()){
            return false;
        }
        return containsHelper(this.root, element);
    }

    private boolean containsHelper(BSTNode<T> current, T element){
        if(current==null){
            return false;
        }

        if(element.compareTo(current.element)==0){
            return true;
        } else if (element.compareTo(current.element)==-1){
            return containsHelper(current.left, element);
        } else{
            return containsHelper(current.right, element);
        }
    }

    public boolean remove(T element){
        BSTNode<T> current = findNode(element);
        BSTNode<T> parent = findParent(element);
        
        if(isEmpty()){
            throw new NullPointerException("Can't remove "+element+", BST is Empty!");
        } else if(element.compareTo(this.root.element)==0){
            throw new NullPointerException("Can't remove "+element+", this is the Root Node!");
        } else if(current.left == null && current.right == null){
            if(current.element.compareTo(parent.element) == -1){
                parent.left = null;
            } else{
                parent.right = null;
            }
            return true;
        } else if(current.left != null && current.right == null){
            if(current.element.compareTo(parent.element)==-1){
                parent.left = current.left;
            } else{
                parent.right = current.left;
            }
            return true;
        } else if (current.left == null && current.right != null){
            if(current.element.compareTo(parent.element)==-1){
                parent.left = current.right;
            } else{
                parent.right = current.right;
            }
            return true;
        } else if (current.left != null && current.right != null){
            T max=maxHelper(current.left);
            
            BSTNode<T> replacement=findNode(max);
            BSTNode<T> replacementParent=findParent(max);
            
            replacementParent.right=replacement.left;
            current.element=replacement.element;
            return true;
        }
        return false;
    }

    private BSTNode<T> findNode(T element){
        if(isEmpty()){
            throw new NullPointerException("Can't remove "+element+", BST is Empty!");
        } 
        return findNodeHelper(this.root, element);
    }

    private BSTNode<T> findNodeHelper(BSTNode<T> current, T element){
        if(element.compareTo(this.root.element)==0){
            throw new NullPointerException("Can't remove "+element+", this is the Root Node!");
        } 

        if(current==null){
            return null;
        } else if(element.compareTo(current.element)==0){
            return current;
        } else if(element.compareTo(current.element)==-1){
            return findNodeHelper(current.left, element);
        } else{
            return findNodeHelper(current.right, element);
        }
    }

    private BSTNode<T> findParent(T element){
        if(isEmpty()){
            throw new NullPointerException("Can't find Parent of "+element+", BST is Empty!");
        } else if(element.compareTo(this.root.element)==0){
            throw new NullPointerException("Can't find Parent of "+element+", its the Root Node!");
        }
        return findParentHelper(this.root, element);
    }

    private BSTNode<T> findParentHelper(BSTNode<T> current, T element){
        if(current == null){
            return null;
        }

        if(element.compareTo(current.element)==-1){
            if(element.compareTo(current.left.element)==0){
                return current;
            } else{
                return findNodeHelper(current.left, element);
            }
        } else{
            if(element.compareTo(current.right.element)==0){
                return current;
            } else{
                return findParentHelper(current.right,element);
            }
        }
    }

}

File: Python/Data Structures & Algorithms Module - Python/Week10/BST Attempt 4/BST.py
"""
BST.py
Billy Cussen
10/12/2020
"""

from BSTNode import BSTNode

class BST:
    def __init__(self):
        self.root = None

    def isEmpty(self):
        return self.root == None

    def size(self):
        if self.isEmpty():
            return 0
        return self.sizeHelper(self.root)

    def sizeHelper(self, current):
        if current == None:
            return 0
        return 1 + self.sizeHelper(current.left)+self.sizeHelper(current.right)

    def insert(self, element):
        if(self.isEmpty()):
            self.root = BSTNode(element)
        else:
            self.insertHelper(self.root, element)

    def insertHelper(self, current, element):
        if ((element > current.element)-(element < current.element)==-1):
            if current.left == None:
                current.left = BSTNode(element)
            else:
                return self.insertHelper(current.left, element)
        else:
            if current.right == None:
                current.right = BSTNode(element)
            else:
                return self.insertHelper(current.right,element)

    def inOrder(self):
        if self.isEmpty():
            print("Unable to do In Order Traversal, BST is Empty!")
        else:
            print("***BST IN ORDER TRAVERSAL***")
            self.inOrderHelper(self.root)
            print("***END***")

    def inOrderHelper(self, current):
        if current == None:
            return
        self.inOrderHelper(current.left)
        print(current.toString())
        self.inOrderHelper(current.right)

    def max(self):
        if(self.isEmpty()):
            return 0
        else:
            return self.maxHelper(self.root)

    def maxHelper(self, current):
        if current==None:
            return None
        if current.right == None:
            return current.element
        else:
            return self.maxHelper(current.right)

    def min(self):
        if(self.isEmpty()):
            return 0
        return self.minHelper(self.root)

    def minHelper(self, current):
        if current == None:
            return None
        else:
            if current.left == None:
                return current.element
            else:
                return self.minHelper(current.left)

    def remove(self, element):
        if self.isEmpty():
            return "Can't remove "+str(element)+", BST is Empty!"
        elif element == self.root.element:
            return "Can't remove "+str(element)+", element is the Root Node"
        
        toRemove = self.findNode(self.root, element)
        parentNode = self.findParent(self.root, element)

        if toRemove.left == None and toRemove.right == None:
            if((toRemove.element > parentNode.element)-(toRemove.element < parentNode.element)==-1):
                parentNode.left = None
            else:
                parentNode.right = None
            return True
        elif toRemove.left != None and toRemove.right == None:
            if((toRemove.element > parentNode.element)-(toRemove.element < parentNode.element)==-1):
                parentNone.left = toRemove.left
            else:
                parentNode.right = toRemove.left
            return True
        elif toRemove.left == None and toRemove.right != None:
            if((toRemove.element > parentNode.element)-(toRemove.element < parentNode.element)==-1):
                parentNode.left = toRemove.right
            else:
                parentNone.right = toRemove.right
            return True
        elif toRemove.left != None and toRemove.right != None:
            maxElement = self.maxHelper(toRemove.left)

            replacement = self.findNode(maxElement, self.root)
            replacementParent = self.findParent(maxElement, self.root)

            replacementParent.right = replacement.left
            toRemove.element = replacement.element
            return True
        return False
        
    def findNode(self, current, element):
        if current == None:
            return None

        if ((element > current.element)-(element < current.element)==0):
            return current
        elif ((element > current.element)-(element < current.element)==-1):
            return self.findNode(current.left, element)
        else:
            return self.findNode(current.right, element)

    def findParent(self, current, element):
        if current == None:
            return None
        elif element == self.root.element:
            return None

        if ((element > current.element)-(element < current.element)==-1):
            if((element > current.left.element)-(element < current.left.element)==0):
                return current
            else:
                return self.findParent(current.left, element)
        else:
            if((element > current.right.element)-(element < current.right.element)==0):
                return current
            else:
                return self.findParent(current.right, element)

            
